---
title: "Basic Bioinformatics in R"
subtitle: "Plotting Graphs and Using BioconductoR"
author: "Ernesto Cuadra Foy"
date: "October 8, 2020"
output:
  xaringan::moon_reader:
    css: xaringan-themer.css
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```
```{r xaringan-themer, include=FALSE, warning=FALSE}
library(xaringanthemer)
style_mono_accent_inverse(
  base_color = "#3C989E",
  header_h1_font_size = '40px',
  header_h2_font_size = '20px',
  header_font_google = google_font("Josefin Sans"),
  text_font_google   = google_font("Montserrat", "300", "300i"),
  code_font_google   = google_font("Fira Mono")
)
```
```{r, include=FALSE}
library(ggpubr)
library(tidyverse)
library(knitr)
library(rmarkdown)
library(rlang)
library(readxl)
library(utils)
library(ggrepel)
library(tidyHeatmap)
library(tidybulk)
```
#Outline
- Bioinformatics in R
- Plotting barplots and determine statistical significance
- Styling a `ggplot` graph ready for publication
- Using `tidybulk` for RNASeq analysis
---
# R and Bioinformatics
- Bioinformatics - like biology - **it's messy**. Dealing with different organisms, different systems, conditions.

- Bioinformatics experiments involve testing and **tweaking parameters** and most tools available involve some sort of coding
  - Commercial tools are expensive and licensed.
  
- R shines with a variety of **statistical tools** available and in generation of **publication-quality** graphs and figures
  - RNA-Seq, population genomics, etc

- Before using any of these tools, make sure to be **working on a tidy and organized dataset**. The tool will not prepare the data for you.

---
# Plotting with `ggplot`
- While `ggplot` is not a tool specific to bioinformatics, it will help you produce quality plots for any kind of dataset you are working with <img src = "https://www.ericchowkokyew.com/wp-content/uploads/2019/07/image-2.png" width=40% border=1 align="right" ><BR CLEAR="left">

- In `ggplot` graphs are made of **layers** that are placed on top of each other 

- Each layer contains **visual objects**. Visual elements are referred to as *geoms* (geometric objects)

- *Geoms* are added to the plot using the **+** sign

- The aesthetic properties help determine the color, shape, and size of the elements for each layer. Referred as *Aes*
![](https://3.bp.blogspot.com/-xG5Jw7-fEpk/WqWjNg49poI/AAAAAAAAHFw/etDhUKvTQIA4UixU_cmUiVC0MEBFuihmQCLcBGAs/s1600/ggplot2_syntax.PNG)

---
# Plotting with `ggplot`
- A Histogram with the `mtcars` dataset
- The `stat` function inside the histogram creates data not from the data itself, however it was mapped to `count`
```{r, echo = TRUE, comment="", message=FALSE, fig.width=5,fig.height=4}
data(mtcars)
ggplot(mtcars, aes(disp))+
  geom_histogram(aes(y=stat(count)))
```
---
# Plotting with `ggplot`
- We can adjust the `stat` function to tell `ggplot` that we want to use calculated aesthetics produced by the statistic
```{r, echo = TRUE, comment="", message=FALSE, fig.width=5,fig.height=4}
data(mtcars)
ggplot(mtcars, aes(disp))+
  geom_histogram(aes(y=stat(count/max(count))))
```
---
# Adding p-values and significance in `ggplot`
- A required package to install is `ggpubr` which provides ggplot2-based publication ready plots
  - Install it as such `install.packages('ggpubr')`
  - Then load it with `library(ggpubr)`
- For this example we will work with the **ToothGrowth** dataset. 
  - Data set contains the result from an experiment studying the effect of vitamin C on tooth growth (**len**) in 60 Guinea pigs
  - Each animal received doses (**supp**) of vitamin C by one of two delivery methods, juice (OJ) or ascorbic acid (VC)
  - Each animal received one of three doses (**dose**) of vitamin C: 0.5, 1, and 2 mg/day
```{r, echo = TRUE, comment="", message=FALSE, fig.width=5,fig.height=4}
data("ToothGrowth")
head(ToothGrowth)
```
---
# Adding p-values and significance in `ggplot`
- `ggpubr` adds two new functions, 
  - `Compare_means()`, easy to use solution to perform one and multiple mean comparisons
  - `stat_compare_means()`, solution to automatically add p-values and significance levels to `ggplot`

- `compare_means(formula, data, method = "wilcox.test", paired = FALSE, group.by = NULL, ref.group = NULL, ...)`
  - In `formula` you have **x**, as a numeric variable and then compare it to a group using **~**
  - In this case the `method` is set by default to be a Wilcoxon test. Allowed variables include `t.test`, `anova`, `kruskal.test`
  - `Paired` indicates that you want a paired test, only used for `t.test` and `wilcox.test`
  
---
# Adding p-values and significance in `ggplot`
- Let's perform a test on our dataset
- **len** contains the length of the Guinea pig teeth and **supp* is the type of delivery method
```{r, echo = TRUE, comment="", message=FALSE, fig.width=5,fig.height=4}
compare_means(len~supp, data = ToothGrowth)
```
  
---
# Adding p-values and significance in `ggplot`
- `stat_compare_means()` acts as a geometrical layer on `ggplot` that performs the same tests but can be used with box plots, dot plots, bar plots and line plots
```{r, echo = TRUE, comment="", message=FALSE, fig.width=5,fig.height=4}
p <- ggboxplot(ToothGrowth, x = 'supp', y= 'len',
               color = 'supp', add = 'jitter')
p
```
---
# Adding p-values and significance in `ggplot`
- Added the p-value on the graph directly
```{r, echo = TRUE, comment="", message=FALSE, fig.width=5,fig.height=4}
p + stat_compare_means(method = "t.test")
```
---
# Adding p-values and significance in `ggplot`
- The `stat_compare_means()` function can be modified further to display the desired aesthetics
- We can see that there is no difference between OJ and VC with regards to tooth length
```{r, echo = TRUE, comment="", message=FALSE, fig.width=5,fig.height=4}
p + stat_compare_means(method = "t.test", aes(label = ..p.signif..),
                       label.x = 1.5, label.y = 40)
```
---
# Adding layers
- We can apply a `facet_grid()` layer which allows us to further stratify our data
- In this case we want to build facets for each dose group, and comparing OJ vs VC
```{r, echo = TRUE, comment="", message=FALSE, fig.width=10,fig.height=4}
p + stat_compare_means(method = "t.test", aes(label = ..p.signif..),
                       label.x = 1.5, label.y = 35)+
                      facet_grid(~dose)
```
---
# Exploring cytokine data from fibroblasts
- Data from "Distinct fibroblast subsets drive inflammation and damage in arthritis", Croft et.al (2019)
- The data is in a format that is easy for us to interpret, but `ggplot` prefers long formats
```{r, echo = TRUE, comment="", message=FALSE, warning=FALSE, fig.width=10,fig.height=4}
data <- read_csv('https://raw.githubusercontent.com/ecuadrafoy/Workshop/master/figure_4a.csv')
head(data)
```
---
# Exploring cytokine data from fibroblasts
- We need to convert from a wide format to a long format, using the function `pivot_long()`
```{r, echo = TRUE, comment="", message=FALSE, warning=FALSE, fig.width=10,fig.height=4}
data_long <- data %>%
  pivot_longer(cols = -Cytokine,
               names_to = c("thy", "status", "repeat"),
               names_sep = "_",
               values_to = "levels") %>% 
  select(-'repeat') %>% 
  unite('status', c('thy','status'))
kable(head(data_long, "simple"))
```
---
# `ggplot` is a highly versatile tool
```{r, echo = TRUE, comment="", warning=FALSE}
my_comparisons <- list( c("Thy1_minus", "Thy1_plus")) #The two groups to compare
 cytokines <- data_long %>% ggplot(aes(x=status, y=levels, fill=status))+ #Pick the axes
   
      stat_summary(fun='mean', geom='bar')+ # Pick the stat to measure, in this case mean
   
      scale_fill_manual(values=c("grey","black"))+ # Pick the colours for your barplot
   
      facet_wrap(~Cytokine, scales = "free_y")+ # Determine the variable by which to create facets
   
      stat_compare_means(comparisons = my_comparisons,
                      method = "t.test", aes(label = ..p.signif..), 
                      size=2, paired = TRUE)+ # Apply the t-test
   
      labs(x='Thy status',
            y = 'Cytokine levels',
            title = 'Figure 4a',
            subtitle = "Croft et al, Nature, 2019")+ #Name the labels
   
      theme(strip.background = element_rect(fill = FALSE))+ # Options to make the configure the graph look
      theme(panel.background = element_rect(fill = "white"))
```
---
# `ggplot` is a highly versatile tool
```{r, echo = TRUE, comment="", message=FALSE, warning=FALSE, fig.width=10,fig.height=6}
cytokines
```
---
# Getting started with `BioconductoR`
- Bioconductor provides tools for the analysis and comprehension of high-throughput genomic data.
- Bioconductor software packages focused on the analysis of single channel Affymetrix and two or more channel cDNA/Oligo microarrays
- As the project has matured, the functional scope of the software packages broadened to include the analysis of all types of genomic data, such as SAGE, sequence, or SNP data.
- As of April 28 2020, there are **1903** packages available
- Bioconductor is a huge module, and has its own installation manager. Be sure to initialize the following line at least once before installing any packages

```{r, echo = TRUE, comment="", eval=FALSE, message=FALSE, warning=FALSE}
if (!requireNamespace("BiocManager"))
    install.packages("BiocManager")
BiocManager::install()
```
---
# Processing RNASeq Data with `tidybulk`
- The tidybulk package provides a tidy data structure and a modular framework for bulk transcriptional analyses
<img src = "https://stemangiola.github.io/tidybulk/inst/logo.png" width=40% align="right" ><BR CLEAR="left">

- `tidybulk` allows differential expression analysis from RNA sequencing datasets in a friendly way.

- The minimal input dataframe for `tidybulk` to work requires the following:
  - Sample column(`chr` or `fctr`) 
  - Transcript/gene column (`chr` or `fctr`)
  - Abundance/counts column (`integer`)

- Creates a tidy transcriptomics object (**tt**) from a tibble (**tbl**) which is a `tidyverse` dataframe

-To install `tidybulk` and any other BioconductoR package, we use the following code:
  - `BiocManager::install("tidybulk")`
  
---
# Processing RNASeq Data with `tidybulk`
- We will be working with the `Pasilla` dataset from Brooks AN et al. 2011, which studied the effect of RNAi knockdown of Pasilla,the Drosophila melanogaster ortholog of mammalian NOVA1 and NOVA2.

- The data provides per-exon and per-gene read counts.

-About **14,000** rows! (Only showing 6 here)

```{r, echo = TRUE, comment="", message=FALSE, warning=FALSE}
pasilla_data <- read_csv('https://raw.githubusercontent.com/ecuadrafoy/Workshop/master/pasilla_gene_counts.csv')
head(pasilla_data)
```
---
# Processing RNASeq Data with `tidybulk`
- We need to clean up this data so that is in a tidy format
```{r, echo = TRUE, comment="", message=FALSE, warning=FALSE}
#Annotation data
coldata = read.csv('https://raw.githubusercontent.com/ecuadrafoy/Workshop/master/pasilla_sample_annotation.csv', row.names = 1)
coldata = coldata[, c("condition", "type")]
#Converting to Long Format and joining annotation table
pasilla_dataset <- 
  pasilla_data %>% as_tibble(rownames = 'transcript') %>%
    pivot_longer(names_to = 'sample',
                 values_to = 'count',
                 cols = -transcript) %>%
    left_join(coldata %>%
                as_tibble(rownames = "sample") %>%
                mutate(sample = gsub("fb", "", sample))) %>%
    mutate_if(is.character, as.factor)
```
---
# Processing RNASeq Data with `tidybulk`
```{r, echo = TRUE, comment="", message=FALSE, warning=FALSE}
head(pasilla_dataset)
```
---
# Processing RNASeq Data with `tidybulk`
-You need to specify which column names contain
  - Sample column(`chr` or `fctr`) 
  - Transcript/gene column (`chr` or `fctr`)
  - Abundance/counts column (`integer`)
- You need to call them in that order
- The end product is a tidybulk-ready dataset called tidy transcriptomic object. 
```{r, echo = TRUE, comment="", message=FALSE, warning=FALSE}
pasilla_tt <- pasilla_dataset %>%  tidybulk(sample, transcript, count)
class(pasilla_dataset)
class(pasilla_tt)
```
---
# Filtering lowly expressed genes
- Now that our data is ready and in a **tt** format, we can start working with it.
- Normalization is performed to eliminate uninteresting differences between samples due to sequencing depth or composition.
- The function `scale_abundance()` generates scaled counts and allows the usage of different normalization methods. By default, `tidybulk` will use trimmed mean of M values (TMM) normalization (Robinson and Oshlack 2010).
- Scaling factors are calculated on filtered transcripts. In this case the `factor_of_interest` is treatment condition
```{r, echo = TRUE, comment="", message=FALSE, warning=FALSE}
pasilla_scaled <- pasilla_tt %>% scale_abundance(factor_of_interest = condition)
pasilla_scaled %>% select(transcript,lowly_abundant) %>% head() #lowly_abundant column is created
```

---
# Principal Component Analysis (PCA)
- With PCA we can reduce the dimensions of data to identify the greatest sources of variation in the data
- This is an example of unsupervised analysis where groups are not specified
- What we hope to see is that the greatest sources of variation in the data are the treatment/groups we are interested in
- It can be a useful tool for quality control and checking for outliers
```{r, echo = TRUE, comment="", message=FALSE, warning=FALSE}
pasilla_scal_PCA <- pasilla_scaled %>%
  reduce_dimensions(method="PCA")
```
---
# Principal Component Aanalysis (PCA)
- This creates the following dataset 
```{r, echo = TRUE, comment="", message=FALSE, warning=FALSE}
pasilla_scal_PCA %>% select(transcript, PC1, PC2) %>% head()
```
---
# Principal Component Aanalysis (PCA)
- `pivot_sample()` returns a table with only sample-related columns, useful for visualization. 
```{r, echo = TRUE, comment="", message=FALSE, warning=FALSE}
pasilla_scal_PCA %>% pivot_sample()
```
---
# Visualizing PCA
- With `pivot_sample()` we can now plot the reduced dimensions. 
- The `ggrepel` package allows you to label dots with names. 
```{r, echo = TRUE, comment="", message=FALSE, warning=FALSE, fig.width=5,fig.height=4}
pasilla_scal_PCA %>%
  pivot_sample() %>%
  ggplot(aes(x=PC1, y=PC2, colour=condition)) +
  geom_point() +
  geom_text_repel(aes(label=sample), show.legend = FALSE) + #From ggrepel package
  theme_bw()
```
---
# Visualizing PCA
- What happens if we include the `type` variable?
- We can see that treatment separates samples on PC1 and that the type of reading is separated by PC2
```{r, echo = TRUE, comment="", message=FALSE, warning=FALSE, fig.width=5,fig.height=4}
pasilla_scal_PCA %>%
  pivot_sample() %>%
  ggplot(aes(x=PC1, y=PC2, colour=condition, shape = type)) +
  geom_point() +
  geom_text_repel(aes(label=sample), show.legend = FALSE) + #From ggrepel package
  theme_bw()
```

---
# Hierarchical clustering with heatmaps
- For heatmaps we use the package `tidyHeatmaps` which relies on the `ComplexHeatmaps` graphical engine. Make sure to install both and activate them before using this package
- `BiocManager::install("ComplexHeatmap")`
- `BiocManager::install("tidyHeatmap")`
```{r, echo = TRUE, eval=FALSE, comment="", message=FALSE, warning=FALSE}
heatmap <- pasilla_scaled %>%
  # filter lowly abundant
  filter(!lowly_abundant) %>%
  # extract 500 most variable genes
  keep_variable( .abundance = count_scaled, top = 500) %>% #keep_variable allows to extract the most variable genes
  # create heatmap
  heatmap(
    .row = transcript,
    .column = sample,
    .value = count_scaled,
    annotation = condition,
  )
```
---
# Hierarchical clustering with heatmaps
```{r, echo = FALSE, comment="", message=FALSE, warning=FALSE, fig.width=10,fig.height=6}
pasilla_scaled %>%
  # filter lowly abundant
  filter(!lowly_abundant) %>%
  # extract 500 most variable genes
  keep_variable( .abundance = count_scaled, top = 500) %>% 
  #keep_variable allows to extract the most variable genes
  
  # create heatmap
  heatmap(
    .row = transcript,
    .column = sample,
    .value = count_scaled,
    annotation = condition,
  )
```
---
#Differential Gene Expression
- After checking the data we can continue testing for differentially expressed genes. 
- The `test_differential_abundance` from tidybulk uses *edgeR* (Robinson, McCarthy, and Smyth 2010) to perform differential expression analysis.
- We take our already prepared tidybulk object and specify what do we want to compare to. This is set after the ~ line.
  - From our PCA analysis we determined that treatment and type of reading were important factors.
- **For some reason I was not able to render this, so this part will be shown in class as well as the MA plot**
```{r, echo = TRUE, eval=FALSE, comment="", message=FALSE, warning=FALSE}
pasilla_de <- pasilla_tt %>% 
  test_differential_abundance(~ condition + type)
```





